### ISSUE TITLE
Warn for use of RawFd or AsRawFd in public interfaces

### ISSUE THREAD

**Issue by asomers** [rust-lang/rust-clippy#10025]
### What it does

RFC 3128 adds `AsFd`, `OwnedFd` and `BorrowedFd`, which are much safer ways to handle file access than `AsRawFd` and `RawFd`.  They're what Rust ideally would've done from the start.  And now that they're in stable Rust there is little reason to use `RawFd` anymore.  I suggest that Clippy warn when safe functions, especially public ones, use `RawFd` or `AsRawFd` in their signatures.

https://rust-lang.github.io/rfcs/3128-io-safety.html

### Lint Name

rawfd_interface

### Category

suspicious

### Advantage

Prevents bugs caused by invalid file descriptors and use-after-close.

### Drawbacks

* Requires a recent MSRV (1.63.0)
* For crates in the middle of the application stack, it may not be possible to convert them to use I/O Safety before lower-level crates do.

### Example

```rust
pub fn do_some_io<FD: AsRawFd>(input: &FD) -> io::Result<()> {
    some_syscall(input.as_raw_fd())
}
```

Could be written as:

```rust
pub fn do_some_io<FD: AsFd>(input: &FD) -> io::Result<()> {
    some_syscall(input.as_fd())
}
```
